### ISSUE TITLE
Provide global sensor dT

### ISSUE THREAD

**Issue by peabody124** [TauLabs/TauLabs#168]
This depends on #163

Right now the dT for each sensor update is computed per sample.  This is going to introduce artificial jitter into the dT estimation.  The PIOS_SENSORS interface provides a natural place to accululate a smoothed estimate of the sampling rate.  This can either be done per sensor or if that is not necessary the fastest sensor (typically gyro) can be the one tracking it.

Suggested API
    PIOS_SENSOR_AcquiredSample() -- this stores a timestamp for smoothing
    PIOS_SENSOR_GetDt() -- returns the smoothed time step

**Comment by mlyle** [comment]
@Trex4Git is looking at this

**Comment by mlyle** [comment]
#1765 did the extent of what we wanted to do for the Endo release-- we can make better with time.

**Comment by Trex4Git** [comment]
I think maybe a strategical decision on sensor dT may be helpful.

In my PR only for the gyro LPF in stabilization a averaged dT is calculated after start (where before a constant fake dT was used). 
All other things there like the PID control an also the LPF the derivative term still use latest dT.
When they should also use a filtered dT I am not sure if only calculating it once after startup would be sufficient, or a continuous filtering of dT is needed.

I have seen that the complementary filter for attitude uses also the same filter (&math with fake dT) for accel and gyro data, as the gyro LPF in stabilization before that PR https://github.com/TauLabs/TauLabs/pull/1765.
So i am thinking about changing it the same way as done for stabilization.

If just adapting we would have than 2 filtered dT calculations, and the suggested API at the beginning of this iusse also looks more like a central place with filtered dT calculations.
The sensors task might be place where PIOS_SENSOR API could be used. Because from there Attitude task and Stabilization task seems to get triggered. 
But as not all tasks there have the same priority, this might be a place for jitter problems (I described there my impression: https://github.com/TauLabs/TauLabs/issues/1786).

So it might be better to have separate filtered dT calculations for attitude and stabilization task, as they might have different delays.
I also seen ideas of running attitude and stabilization loops at different rates. That would also be a reason to use separate dT, I think.

**Comment by mlyle** [comment]
I think we should have a single global dT.

This is because a lot of the subsequent uses of dT really have to do with the duration the sensor data is interpreted over.  And this doesn't have so much to do with when we're looking at the information but instead when it's measured.

I know we may eventually run portions of stabilization at a different rate, but it's likely to be an integer ratio to the attitude rate, so it's possible to multiply appropriately (It's likely to always be 1/1 or 1/2).

**Comment by Trex4Git** [comment]
I also think a global dT is nicer (and had already plans to maybe integrate that in https://github.com/TauLabs/TauLabs/pull/1765), but after reading some ideas on IRC in was not sure what future plans are there.
So i just have quick written my impressions.

Is no one else is interested in doing that, i can work on this as already having put some thoughts on it.
Not sure if this would be in the scope of next release, and how fast a implementation should be finished. But in principal it should be much work.
It just has to be decided if all control parts in stabilization and attitude which use a dT should use the global, filtered dT in future?

**Comment by mlyle** [comment]
Yes, I think everything should use the global, filtered dT (multiplying it as appropriate if portions of the loop run at a different rate).

I don't think it will make the next release, but the work can wait in the pull queue until we're ready to start on that branch.
